src/radiation/radiation_system.hpp

```diff
@@ -2006,4 +1934,524 @@ void RadSystem<problem_t>::AddSourceTerms(array_t &consVar, arrayconst_t &radEne
+				double deltaR = NAN;
+				double F_D = NAN;
+
+				const double resid_tol = 1.0e-11; // 1.0e-15;
+				const int maxIter = enable_dust_gas_thermal_coupling_model_ ? 100 : 50;
+				int n = 0;
+				for (; n < maxIter; ++n) {
+					T_gas = quokka::EOS<problem_t>::ComputeTgasFromEint(rho, Egas_guess, massScalars);
+					AMREX_ASSERT(T_gas >= 0.);
+
+					// dust temperature
+					if constexpr (!enable_dust_gas_thermal_coupling_model_) {
+						T_d = T_gas;
+					} else {
+						if (n == 0) {
+							const quokka::valarray<double, 1> Erad_guess_vec{Erad_guess};
+							T_d = ComputeDustTemperature(T_gas, T_gas, rho, Erad_guess_vec, dustGasCoeff_local);
+							AMREX_ASSERT_WITH_MESSAGE(T_d >= 0., "Dust temperature is negative!");
+						} else {
+							const auto Lambda_gd = R / (dt * chat / c);
+							T_d = T_gas - Lambda_gd / (dustGasCoeff_local * num_den * num_den * std::sqrt(T_gas));
+							AMREX_ASSERT_WITH_MESSAGE(T_d >= 0., "Dust temperature is negative!");
+						}
+						if (T_d < 0.0) {
+							amrex::Gpu::Atomic::Add(p_num_failed_dust_local, 1);
+						}
+					}
+
+					fourPiBoverC = ComputeThermalRadiationSingleGroup(T_d);
+
+					kappaP = ComputePlanckOpacity(rho, T_d);
+					kappaE = ComputeEnergyMeanOpacity(rho, T_d);
+					AMREX_ASSERT(!std::isnan(kappaP));
+					AMREX_ASSERT(!std::isnan(kappaE));
+					AMREX_ASSERT(kappaP >= 0.0);
+					AMREX_ASSERT(kappaE >= 0.0);
+					if (kappaE > 0.0) {
+						kappaPoverE = kappaP / kappaE;
+					} else {
+						kappaPoverE = 1.0;
+					}
+
+					// In the first loop, calculate kappaF, work, tau0, R
+					if (n == 0) {
+						kappaF = ComputeFluxMeanOpacity(rho, T_d);
+						AMREX_ASSERT(!std::isnan(kappaF));
+
+						if constexpr ((beta_order_ != 0) && (include_work_term_in_source)) {
+							// compute the work term at the old state
+							if (ite == 0) {
+								const double frad0 = consPrev(i, j, k, x1RadFlux_index);
+								const double frad1 = consPrev(i, j, k, x2RadFlux_index);
+								const double frad2 = consPrev(i, j, k, x3RadFlux_index);
+								// work = v * F * chi
+								work = (x1GasMom0 * frad0 + x2GasMom0 * frad1 + x3GasMom0 * frad2) * (2.0 * kappaE - kappaF) *
+								       chat / (c * c) * lorentz_factor_v * dt;
+							}
+						}
+
+						tau0 = dt * rho * kappaP * chat * lorentz_factor;
+						tau = tau0;
+						R = (fourPiBoverC - Erad_guess / kappaPoverE) * tau0 + work;
+						// tau0 is used as a scaling factor for Rvec
+						tau0 = std::max(tau0, 1.0);
+					} else { // in the second and later loops, calculate tau and E (given R)
+						tau = dt * rho * kappaP * chat * lorentz_factor;
+						if (tau > 0.0) {
+							Erad_guess = kappaPoverE * (fourPiBoverC - (R - work) / tau);
+							if constexpr (force_rad_floor_in_iteration) {
+								if (Erad_guess <= 0.0) {
+									Egas_guess -= (c / chat) * (Erad_floor_ - Erad_guess);
+									Erad_guess = Erad_floor_;
+								}
+							} else {
+								// AMREX_ASSERT(Erad_guess > 0.0);
```

Comment:
Is there a reason to keep this `else` statement?

Reply:
No. I have deleted it.